packages/labs/virtualizer/src/layouts/shared/BaseLayout.ts

```diff
@@ -453,51 +426,49 @@ export abstract class BaseLayout<C extends BaseLayoutConfig> implements Layout {
     } as ScrollToOptions;
   }
 
-  private _emitUnpinned() {
-    this.dispatchEvent(new CustomEvent('unpinned'));
+  private _sendUnpinnedMessage() {
+    this._hostSink({
+      type: 'unpinned',
+    });
   }
 
-  protected _emitRange() {
-    const detail: InternalRange = {
-      first: this._first,
-      last: this._last,
+  private _sendVisibilityChangedMessage() {
+    this._hostSink({
+      type: 'visibilityChanged',
       firstVisible: this._firstVisible,
       lastVisible: this._lastVisible,
-    };
-    this.dispatchEvent(new CustomEvent('rangechange', {detail}));
+    });
   }
 
-  protected _emitScrollSize() {
-    const detail = {
-      [this._sizeDim]: this._scrollSize,
-      [this._secondarySizeDim]: null,
+  protected _sendStateChangedMessage() {
+    const childPositions: ChildPositions = new Map();
+    if (this._first !== -1 && this._last !== -1) {
+      for (let idx = this._first; idx <= this._last; idx++) {
+        childPositions.set(idx, this._getItemPosition(idx));
+      }
+    }
+    const message: StateChangedMessage = {
+      type: 'stateChanged',
+      scrollSize: {
+        [this._sizeDim]: this._scrollSize,
+        [this._secondarySizeDim]: null,
+      } as Size,
```

Comment:
tiny nit: it seems the virtualiser actually handles nullish dimensions. is it worth updating `Size` to reflect that and drop the cast here?

in the virtualiser it already checks `size.width !== null` etc even though the type says it can't ever be `null`

near enough a non-issue but maybe still worth me pointing out

Reply:
This is a good observation; however, I think it might be a longer thread than we want to pull on as part of this PR. The `Size` type is used in quite a few places in the Virtualizer code base, and I am quite certain that in some of them the logic doesn't anticipate or guard against `null` values. My inclination is to take care of this independently of this PR, perhaps as part of a more systematic review of internally used types. @usergenic, thoughts?

Reply:
@graynorton @43081j I'm all for considering making Size nullable as separate PR.  It will give time to fully vet the change; there's enough going on here already.

Reply:
Yup of course. Didn't mean to distract at all. Was mostly an observation 👍